azureDatabaseUtils/src/main/java/bio/terra/workspace/azureDatabaseUtils/database/DatabaseService.java

```diff
@@ -80,4 +90,157 @@ public void restoreNamespaceRoleAccess(String namespaceRole) {
+    // workspace at once;
+    // (2) the workspace identity can grant itself access at any time, so revoking the role
+    // doesn't protect us.
+    databaseDao.grantRole(adminUser, dbName);
+
+    List<String> commandList = generateCommandList("pg_dump", dbName, dbHost, dbPort, adminUser);
+    Map<String, String> envVars = null;
+    try {
+      envVars = Map.of("PGPASSWORD", determinePassword());
+    } catch (PSQLException e) {
+      logger.error(e.getMessage());
+    }
+
+    logger.info("running DatabaseService.pgDump: {}", String.join(" ", commandList));
+    logger.info("blobContainerName: {}", blobContainerName);
+
+    LocalProcessLauncher localProcessLauncher = new LocalProcessLauncher();
+    localProcessLauncher.launchProcess(commandList, envVars);
+
+    storage.streamOutputToBlobStorage(
+        localProcessLauncher.getInputStream(),
+        blobFileName,
+        blobContainerName,
+        blobContainerUrlAuthenticated);
+
+    checkForError(localProcessLauncher);
+  }
+
+  public void pgRestore(
+      String dbName,
+      String dbHost,
+      String dbPort,
+      String adminUser,
+      String blobFileName,
+      String blobContainerName,
+      String blobContainerUrlAuthenticated) {
+
+    boolean doCleanup = false;
+
+    // Grant the database role (dbName) to the workspace identity (adminUser).
+    // In theory, we should be revoking this role after the operation is complete.
+    // We are choosing to *not* revoke this role for now, because:
+    // (1) we could run into concurrency issues if multiple users attempt to clone the same
+    // workspace at once;
+    // (2) the workspace identity can grant itself access at any time, so revoking the role
+    // doesn't protect us.
+    try {
+      databaseDao.grantRole(adminUser, dbName);
+
+      List<String> commandList = generateCommandList("psql", dbName, dbHost, dbPort, adminUser);
+      Map<String, String> envVars = null;
+      try {
+        envVars = Map.of("PGPASSWORD", determinePassword());
+      } catch (PSQLException e) {
+        logger.error(e.getMessage());
+      }
+      LocalProcessLauncher localProcessLauncher = new LocalProcessLauncher();
+      localProcessLauncher.launchProcess(commandList, envVars);
+
+      logger.info("running DatabaseService.pgRestore: {}", String.join(" ", commandList));
+      logger.info("blobContainerName: {}", blobContainerName);
+
+      storage.streamInputFromBlobStorage(
+          localProcessLauncher.getOutputStream(),
+          blobFileName,
+          blobContainerName,
+          blobContainerUrlAuthenticated);
+
+      // doCleanup isn't set to true until checkForError completes without throwing an exception
+      checkForError(localProcessLauncher);
+      databaseDao.reassignOwner(adminUser, dbName);
+      doCleanup = true;
+    } finally {
+      if (doCleanup) {
+        storage.deleteBlob(blobFileName, blobContainerName, blobContainerUrlAuthenticated);
```

Comment:
I don't think cleanup should be in this utility. I think it should be a step in the stairway flight after the step that calls pg restore

Reply:
I'd argue to just do it here (since we wont need it for anything else, might as well keep it all contained in one place) if we do this somewhere else - aren't we are more likely to end up with blobs that may not get cleaned up?

Reply:
(sorry for my slow reply...)

> aren't we are more likely to end up with blobs that may not get cleaned up?

maybe, I'm not sure. but doing the cleanup in a separate step fits the stairway model better, because each step needs to be idempotent. (If this step completes successfully, it'll delete the dump file, making subsequent runs fail)